pkg/kv/kvserver/replica_application_result.go

```diff
@@ -104,11 +104,40 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
 		case kvserverbase.ProposalRejectionPermanent:
 			cmd.response.Err = pErr
 		case kvserverbase.ProposalRejectionIllegalLeaseIndex:
-			// If we failed to apply at the right lease index, try again with a
-			// new one. This is important for pipelined writes, since they don't
-			// have a client watching to retry, so a failure to eventually apply
-			// the proposal would be a user-visible error.
-			pErr = tryReproposeWithNewLeaseIndex(ctx, cmd, (*replicaReproposer)(r))
+			// Reset the error as it's now going to be determined by the outcome of
+			// reproposing (or not).
+			//
+			// Note that if pErr remains nil, we will mark the proposal as non-local at
+			// the end of this block and return, so we're not hitting an NPE near the end
+			// of this method where we're attempting to reach into `cmd.proposal`.
+			//
+			// This control flow is sketchy but it preserves existing behavior
+			// that would be too risky to change at the time of writing.
+			//
+			pErr = nil
+			// If the command associated with this rejected raft entry already applied
+			// then we don't want to repropose it. Doing so could lead to duplicate
+			// application of the same proposal. (We can see hit this case if an application
+			// batch contains multiple copies of the same proposal, in which case they are
+			// all marked as local, the first one will apply (and set p.applied) and the
+			// remaining copies will hit this branch).
+			//
+			// Similarly, if the proposal associated with this rejected raft entry is
+			// superseded by a different (larger) MaxLeaseIndex than the one we decoded
+			// from the entry itself, the command must have already passed through
+			// tryReproposeWithNewLeaseIndex previously (this can happen if there are
+			// multiple copies of the command in the logs; see
+			// TestReplicaRefreshMultiple). We must not create multiple copies with
+			// multiple lease indexes, so don't repropose it again. This ensures that at
+			// any time, there is only up to a single lease index that has a chance of
+			// succeeding in the Raft log for a given command.
+			if !cmd.proposal.applied && !cmd.proposal.Supersedes(cmd.Cmd.MaxLeaseIndex) {
```

Comment:
This inverts the condition in the (now removed) early return in `tryReproposeWithNewLeaseIndex`.